shared/util/codeql/util/FilePath.qll

```diff
@@ -0,0 +1,74 @@
+/** Provides a utility for normalizing filepaths. */
+
+/**
+ * A filepath that should be normalized.
+ * Extend to provide additional strings that should be normalized as filepaths.
+ */
+abstract class NormalizableFilepath extends string {
+  bindingset[this]
+  NormalizableFilepath() { any() }
+
+  private string getComponent(int i) { result = this.splitAt("/", i) }
+
+  private int getNumComponents() { result = strictcount(int i | exists(this.getComponent(i))) }
+
+  /** count .. segments in prefix in normalization from index i */
```

Comment:
I see you simply copied my algorithm suggestion verbatim, but these qldocs could be a bit nicer now that the code is moving from a slack thread and into actual main.